content/blog/2019-09-25-bril-to-llvm-ocaml.md

```diff
@@ -0,0 +1,151 @@
+
+### LLVM Code Generation
+
+LLVM code generation consisted of 2 main parts. First, since LLVM IR requires each basic block to have a terminator, I decided first generate all of the basic blocks of a Bril program. I could have probably simply looked for all labels with no preceding terminator in a Bril program and added a jump, but since I already written the code for creating and processing basic blocks, I decided to generate LLVM code at a basic block level.
+
+Next, since I didn't implement an SSA transformation for Bril and because Bril variables can be overwritten, I need to create some "stack" space for all of the variables. The way I did this was by using the `alloca` LLVM instruction at the top of the function. I first collected all of the variables that were written to in a function and mapped them to an index in the stack. I also decided to make all variables the LLVM `i64` type. OCaml ints are 63 bits and I wanted to support the largest range of numbers possible. So at the beginning of a function, I inserted a call to `alloca i64, i64 n`,where `n` was the number of unique variables written to. Note, that if a variable is used without being defined anywhere in the function, this will generate a blank instruction and likely cause LLVM to fail when it typechecks the generated code.
+
+Whenever a variable is used as part of an operation it is loaded from the stack. Whenever a variable is modified by an operation, the result of the operation is written back to the variable's location on the stack. So, for example, an add of two variables first performs a load of both variables from the stack into fresh variable names, adds the two variables together with an LLVM `add` instruction, and stores the result back to the `dest` variable of the Bril add instruction.
+
+One other interesting aspect of this project was implementing the print function. I wrote some C code in a `helpers.c` file that just defined a `printi` and a `printb` function that printed out a 64-bit integer and a boolean `true` or `false` respectively. This C code was then compiled and linked in with the generated `.ll` file to create the final binary. In order to figure out which function to generate for each print operation, I added some code to get the type of a variable when it was defined. For example, if I saw:
+
+```
+v: int = add v w;
+```
+
+in a Bril program, I would say that variable `v` has type `int`. This may not work in general because the type of a Bril variable can technically be dynamic. For example, this is legal Bril:
```

Comment:
True, but FWIW we have discussed on Slack the idea of changing the spec to reclassify this as "illegal." That would probably be a good idea.